### PR TITLE
fix: do not eliminate possible out of bounds errors

### LINKED ISSUE #5296: Dead Instruction Elimination pass removes out of bounds array sets
### Aim

While moving some tests from constrained to `unconstrained` so they run faster (we're only checking function logic and assume circuit generation is fine), the code that used to work started failing with "Array index out of bounds" errors.

### Expected Behavior

Behavior should be consistent between constrained and unconstrained contexts.

### Bug

Out of bounds errors can be masked in certain scenarios, specifically when copying array positions from one to another. However, illegal acceses are always detected as expected in unconstrained functions.

### To Reproduce

```rust 
// Taken from noir-protocol-circuits
fn arr_copy_slice<T, N, M>(src: [T; N], mut dst: [T; M], offset: u32) -> [T; M] {
    for i in 0..dst.len() {
        // We are illegally accessing the src array here if dst.len() > src.len()
        dst[i] = src[i + offset];
    }
    dst
}

unconstrained fn boom<N, M>(src: [Field; N], dst: [Field; M]) -> [Field; M] {
    arr_copy_slice(src, dst, 0)
}


fn main() {
    // This shouldn't work
    let small_array_src = [1; 5];
    let mut big_array_dst = [0; 6];
    let mut result = arr_copy_slice(small_array_src, big_array_dst, 0);

    // Uncomment for fireworks. This *always* get caught in unconstrained land
    //let result = boom();

    // This is fine
    let first = result[0];
    dep::std::println(first);
    // This is also fine, since we are overwriting the "illegal" position.
    // BUT if we comment this line out, fireworks again (since we are now "using" the illegal position)
    result[5] = 1;

    dep::std::println(result[5]);
}
```

### Project Impact

Nice-to-have

### Impact Context

_No response_

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Installation Method

Compiled from source

### Nargo Version

_No response_

### NoirJS Version

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #5296

## Summary

Is there another way to do this? That is, can we not remove these instructions when we don't know if they are out of bounds? 

## Additional Context

None.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs [hunk 1]
```diff
@@ -334,6 +334,33 @@ impl Instruction {
...
+                        known_index < array_length.into()
+                    } else {
+                        // If the index is not known at compile-time we can't remove this instruction as this
+                        // might be an index out of bounds.
+                        false
+                    }
+                } else {
+                    // TODO: the underlying type is a Slice. We consider this instruction can be removed,
+                    // though in practice this can also produce an index out of bounds.
+                    true
```
[the remark this answers is not in the document]
Comment (asterite) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Sounds good! I think I pushed another commit for this before seeing your comment.

**Comment by asterite** [comment]
> I am a bit wary of this PR though. Since it seems like a potential performance degradation in larger programs and on slices. E.g. programs which make use of more dynamic indexing or more slices will be hit harder by this.

Agreed. I'll take a look at the regressions to see if there are cases where they can still be removed in case the index is not known.

**Comment by jfecher** [comment]
I wonder if it'd help constraint counts at all if we treated all `array_get` and `array_set` instructions as unchecked and during ssa-gen lower each get & set with an `assert(i < array.len())` instruction before it. That way the array get & set can be treated as pure and only the assert before it would have a side effect.

**Comment by asterite** [comment]
Good idea! I was looking at the `slice_dynamic_index` program and kept only a portion of it:

```rust
fn main(x: Field) {
    let mut slice = &[0, 1, 2, 3, 4];
    let _ = slice[x - 1];
}
```

The SSA before this PR is this:

```
acir(inline) fn main f0 {
  b0(v0: Field):
    v24 = sub v0, Field 1
    v25 = cast v24 as u32
    v26 = lt v25, u32 5
    constrain v26 == u1 1 '"Index out of bounds"'
    return
}
```

The new SSA is this:

```
acir(inline) fn main f0 {
  b0(v0: Field):
    v24 = sub v0, Field 1
    v25 = cast v24 as u32
    v26 = lt v25, u32 5
    constrain v26 == u1 1 '"Index out of bounds"'
    v27 = array_get [Field 0, Field 1, Field 2, Field 3, Field 4], index v25
    return
}
```

So two things:
1. It seems we already insert a constrain to see if we are in-bounds
2. here `v25` will always be in-bounds because of the previous constrain, so the next `array_get` could be removed

What's interesting is that this program has a different SSA:

```rust
fn main(x: Field) {
    let mut slice = &[0, 1, 2, 3, 4];
    slice[x - 1] = 1;
}
```

The SSA before this PR is this one (empty!):

```rust
acir(inline) fn main f0 {
  b0(v0: Field):
    return
}
```

In this PR it's this:

```
After Array Set Optimizations:
acir(inline) fn main f0 {
  b0(v0: Field):
    v23 = sub v0, Field 1
    v24 = cast v23 as u32
    v25 = array_set mut [Field 0, Field 1, Field 2, Field 3, Field 4], index v24, value Field 1
    return
}
```

I was looking at our code and there's `codegen_slice_access_check` but (according to the above SSA's) it seems we don't do that check for set, only for get. Maybe that's the main issue? (I'll check)

**Comment by asterite** [comment]
However... that constrain only seems to be inserted for slices, not for arrays... not sure why.

**Comment by vezenovm** [comment]
> However... that constrain only seems to be inserted for slices, not for arrays... not sure why.

Discussed on a call, but posting here visibility as well.

We currently only code gen a [length constrain on slices](https://github.com/noir-lang/noir/blob/28211a397b810c661204b45b7da06f7cad345278/compiler/noirc_evaluator/src/ssa/ssa_gen/mod.rs#L421) due to their lengths being dynamic. 

I see a couple options:
1. We insert these assertions for array accesses, not just slices.
2. We do not eliminate possible OOB errors as this PR currently does. 

I lean towards option 2 as although it may cause some larger programs, it should only cause larger programs where we have unused array gets/sets. I think it is ok to say that a developer should remove these themselves. We can add an unused variable warning array gets/sets that are later unused, and then in the long-term perhaps we can have a "release" mode that removes these automatically.

**Comment by jfecher** [comment]
> I lean towards option 2 as although it may cause some larger programs, it should only cause larger programs where we have unused array gets/sets. 

I think unused gets & sets are more widespread than you think since in the 3 slices tests that increased in size in the CI this leads to a 3-5% increase in constraint counts which seems significant.

**Comment by asterite** [comment]
Just a note: originally the PR kept `array_get` on slices if the index might be out of bounds. Then I noticed those already have a bounds check, so now those are removed (like before). However, I didn't see another "Changes to circuit sizes" comment. I think the diff now should be smaller (though still not the same as before). Is there a way to force that comment to be produced?

**Comment by jfecher** [comment]
@asterite it is automatically updated with each commit you make as part of the "Report gates diff" CI action

**Comment by asterite** [comment]
To summarize all the scenarios... here are four scenarios and whether they correctly give you "index out of bounds".

## Array get

### Known index out of bounds

```rust
fn main() {
    let array = [1, 2, 3];
    let _ = array[10];
}
```

Expected: index out of bounds
Actual: no error ❌

### Unknown index

```rust
fn main(x: Field) {
    let array = [1, 2, 3];
    let _ = array[x]; // x is 10 in Prover.toml
}
```

Expected: index out of bounds
Actual: no error ❌

## Array set

### Known index out of bounds

```rust
fn main() {
    let mut array = [1, 2, 3];
    array[10] = 42;
}
```

Expected: index out of bounds
Actual: no error ❌

### Unknown index

```rust
fn main(x: Field) {
    let mut array = [1, 2, 3];
    array[x] = 42; // x is 10 in Prover.toml
}
```

Expected: index out of bounds
Actual: no error ❌

## Slice get

### Known index out of bounds

```rust
fn main() {
    let slice = &[1, 2, 3];
    let _ = slice[10];
}
```

Expected: index out of bounds
Actual: index out of bounds ✅ 

### Unknown index

```rust
fn main(x: Field) {
    let slice = &[1, 2, 3];
    let _ = slice[x]; // x is 10 in Prover.toml
}
```

Expected: index out of bounds
Actual: index out of bounds ✅ 

## Slice set

### Known index out of bounds

```rust
fn main() {
    let mut slice = &[1, 2, 3];
    slice[10] = 42;
}
```

Expected: index out of bounds
Actual: no error ❌

### Unknown index

```rust
fn main(x: Field) {
    let mut slice = &[1, 2, 3];
    slice[x] = 42; // x is 10 in Prover.toml
}
```

Expected: index out of bounds
Actual: no error ❌

# Conclusion

I'm not sure how to proceed here. It seems a lot of checks are missing but I don't know if they should go into SSA or acir-gen.

**Comment by asterite** [comment]
Closed in favor of #5676

**Comment by asterite** [comment]
Closing because we chose #5691 instead.

**Pull request closed without merging**